FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs

```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
                 RowKey = TableKeyProvider.LeaseRowKey(_queueType),
                 WorkingInstanceGuid = _instanceGuid,
                 HeartbeatDateTime = DateTimeOffset.UtcNow,
             };
 
-            // add the initial trigger entity to table
-            await _metadataStore.AddEntityAsync(initialTriggerLeaseEntity, cancellationToken);
+            bool isSucceeded = await _metadataStore.TryAddEntityAsync(initialTriggerLeaseEntity, cancellationToken);
+
+            _logger.LogInformation(isSucceeded
+                ? "Create initial trigger lease entity successfully."
+                : "Failed to create initial trigger lease entity, the entity already exists.");
 
-            _logger.LogInformation("Create initial trigger lease entity successfully.");
+            return isSucceeded;
         }
 
-        private async Task<bool> TryPullAndProcessTriggerEntity(CancellationTokenSource cancellationTokenSource)
+        /// <summary>
+        /// a while-loop to check and update trigger entity
+        /// </summary>
+        /// <param name="cancellationToken">cancellation token</param>
+        /// <returns>return true if The job has been scheduled to end, otherwise return false</returns>
+        private async Task<bool> LoopToCheckAndUpdateTriggerEntityAsync(CancellationToken cancellationToken)
         {
-            var cancellationToken = cancellationTokenSource.Token;
-            while (!cancellationToken.IsCancellationRequested)
-            {
-                var intervalDelayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), CancellationToken.None);
+            _logger.LogInformation($"scheduler instance {_instanceGuid} starts to check and update trigger entity in loop.");
 
-                var currentTriggerEntity = await _metadataStore.GetCurrentTriggerEntityAsync(_queueType, cancellationToken) ??
-                                           await CreateInitialTriggerEntity(cancellationToken);
+            bool scheduledToEnd = false;
+            while (!scheduledToEnd && !cancellationToken.IsCancellationRequested)
+            {
+                Task intervalDelayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), cancellationToken).ContinueWith(_ => { });
 
-                // if current trigger entity is still null after initializing, just skip processing it this time and try again next time.
-                // this case should not happen.
-                if (currentTriggerEntity != null)
+                try
                 {
-                    switch (currentTriggerEntity.TriggerStatus)
+                    scheduledToEnd = await CheckAndUpdateTriggerEntityAsync(cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    // if cancelled, the exception will be caught here and exit while loop as cancellationToken.IsCancellationRequested is true
```

Comment:
Seems not exit?

Reply:
Yes, if it isn't a cancellation exception, will not exit and retry later.